src/storage/core.rs

```diff
@@ -0,0 +1,471 @@
+                .config
+                .work_dir
+                .as_ref()
+                .ok_or(Error::Unitialized)?,
+        );
+        if path.exists() {
+            debug!("work dir exists: {}", path.display());
+        } else {
+            debug!("creating work dir recursively: {}", path.display());
+            fs::create_dir_all(path).map_err(Error::IO)?;
+        }
+
+        let lock_file_path = path.join(LOCK_FILE);
+        if lock_file_path.exists() {
+            return Err(Error::WorkDirInUse);
+        }
+
+        debug!("try to open lock file: {}", lock_file_path.display());
+        let lock_file = OpenOptions::new()
+            .create_new(true)
+            .write(true)
+            .open(&lock_file_path)
+            .map_err(Error::IO)?;
+        await!(self.inner.lock()).lock_file = Some(lock_file);
+        Ok(())
+    }
+
+    async fn init_new(&mut self) -> Result<()> {
+        let mut inner = await!(self.inner.lock());
+        let next = self.next_blob_name()?;
+        inner.active_blob = Some(Blob::open_new(next).map_err(Error::InitActiveBlob)?.boxed());
+        debug!(
+            "created new active blob: {}",
+            inner.active_blob.as_ref().unwrap().id()
+        );
+        Ok(())
+    }
+
+    async fn init_from_existing(&mut self, files: Vec<DirEntry>) -> Result<()> {
+        let dir_content = files.iter().map(DirEntry::path);
+        let dir_files = dir_content.filter(|path| path.is_file());
+        let mut temp_blobs: Vec<_> = dir_files
+            .filter(|path| {
+                path.extension()
+                    .map(|os_str| os_str.to_str().unwrap())
+                    .unwrap_or("")
+                    == BLOB_FILE_EXTENSION
+            })
+            .filter_map(|path| {
+                let blob = Blob::from_file(path.clone()).ok()?;
+                blob.check_data_consistency()
+                    .map_err(|e| error!("Check data consistency failed: {:?}", e))
+                    .ok()?;
+                Some(blob)
+            })
+            .collect();
+        temp_blobs.sort_by_key(Blob::id);
+        let active_blob = temp_blobs.pop().unwrap().boxed();
+        let mut inner = await!(self.inner.lock());
+        inner.active_blob = Some(active_blob);
+        inner.blobs = temp_blobs;
+        self.shared.next_blob_id.store(
+            inner.max_id().map(|i| i + 1).unwrap_or(0),
+            Ordering::Relaxed,
+        );
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct Inner {
+    active_blob: Option<Box<Blob>>,
+    blobs: Vec<Blob>,
+    lock_file: Option<File>,
+}
+
+impl Inner {
+    fn new() -> Self {
+        Self {
+            active_blob: None,
+            blobs: Vec::new(),
+            lock_file: None,
+        }
+    }
+
+    fn max_id(&self) -> Option<usize> {
+        let active_blob_id = self.active_blob.as_ref().map(|blob| blob.id());
+        let blobs_max_id = self.blobs.iter().max_by_key(|blob| blob.id()).map(Blob::id);
```

Comment:
As we store blobs as sorted array - this could be simplified as blobs[blobs.len()-1].id

Reply:
Done